### ISSUE TITLE
Evaluating multiple scores

### ISSUE THREAD

**Issue by shaddyab** [scikit-optimize/scikit-optimize#797]
Is it possible to run  BayesSearchCV with multiple scores?
I am getting the following error when inputting a dictionary of scores

```
    297         raise ValueError("For evaluating multiple scores, use "
    298                          "sklearn.model_selection.cross_validate instead. "
--> 299                          "{0} was passed.".format(scoring))
    300     else:
    301         raise ValueError("scoring value should either be a callable, string or"

ValueError: For evaluating multiple scores, use sklearn.model_selection.cross_validate instead. {'neg_mean_squared_error': 'neg_mean_squared_error', 'neg_mean_absolute_error': 'neg_mean_absolute_error'} was passed.
```

**Comment by chnzhangrui** [comment]
Hi,

Can you illustrate what do you want to do? Maybe a snippet of your code?

**Comment by shaddyab** [comment]
I am trying to run a cross validation Bayes search where each fold is evaluated on multiple metrics (as you can see in the example below, this can be done using sklearn's GridSearchCV function)

```
from skopt import BayesSearchCV
from sklearn.model_selection import GridSearchCV
from sklearn.datasets import make_hastie_10_2
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import make_scorer
from sklearn.metrics import accuracy_score

X, y = make_hastie_10_2(n_samples=8000, random_state=42)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.75, test_size=.25, random_state=0)
```
The following code will run with out any errors:

```
gs = GridSearchCV(
    SVC(),
    param_grid={'kernel': ['linear', 'poly', 'rbf']},
    cv=2,
    scoring={'AUC': 'roc_auc', 'Accuracy': make_scorer(accuracy_score)},
    refit= False
)

gs.fit(X_train, y_train)

```

However, I get an error when running the following code:

```

opt = BayesSearchCV(
    SVC(),
    {
        'kernel': ['linear', 'poly', 'rbf'],  # categorical parameter
    },
    n_iter=2,
    cv=2,
    scoring={'AUC': 'roc_auc', 'Accuracy': make_scorer(accuracy_score)},
    refit= False
)

opt.fit(X_train, y_train)

```

Error message:

>     428         raise ValueError("For evaluating multiple scores, use "
>     429                          "sklearn.model_selection.cross_validate instead. "
> --> 430                          "{0} was passed.".format(scoring))
>     431     else:
>     432         raise ValueError("scoring value should either be a callable, string or"
> 
> ValueError: For evaluating multiple scores, use sklearn.model_selection.cross_validate instead. {'AUC': 'roc_auc', 'Accuracy': make_scorer(accuracy_score)} was passed.

**Comment by chnzhangrui** [comment]
Hi,

Thanks for the clarification! The current version of `skopt=0.5.2` doesn't support multiple scores. I think it is because the Bayesian approach needs one score to make predictions and multiple scores will confuse it. Why do you want multiple scores for Bayesian search?

**Comment by shaddyab** [comment]
I am trying to check how are different metrics are affected by the selected hyperparametsr. For example, accuracy and F1.

**Comment by chnzhangrui** [comment]
Hi,

In that case, running separate jobs with different metrics might be helpful.

**Comment by shaddyab** [comment]
Can you please provide an example of running multiple jobs?

Do you mean running multiple optimization (aka codes) in series? if so, then this appears to be too time consuming. I am not trying to run multiple optimization on multiple metrics, but rather a single optimization on a single metric where the validation dataset is evaluated on multiple metrics.

**Comment by chnzhangrui** [comment]
Yeah, I literally mean running multiple optimizations (aka codes) with different scores since I thought you want to compare the impact of different scores. I don't have example codes though.

**Comment by qiagu** [comment]
I vote for the multiple scoring feature, just like in scikit-learn GridSearchCV. There can be a primary scorer for optimization, other scorers for reporting only.

**Comment by bbernst** [comment]
Any updates on this?

**Comment by pietermarsman** [comment]
For some time now I want to contribute something to scikit-optimize. This seems like an issue that adds value, but does not needs profound knowledge on all the intricate details of scikit-optimize. So I expect that I can pull this off ;) 

@holgern @chnzhangrui are you involved in the development of scikit-optimize? Can you give me some clues/suggestions/recommendations on how to approach this?

**Comment by chnzhangrui** [comment]
Hi,

@holgern is the main developer and maintainer right now. Maybe he will have some suggestions on how to implement this feature.

**Comment by AntonBiryukovUofC** [comment]
That does look like a helpful feature. Could one implement it via MonitoringCallbacks ,similar to how Keras does it ?

**Comment by jmahenriques** [comment]
I'm actually OK with using a single score to drive the search, but it would be nice to be able to recover the predictions at each fold and iteration so that one could calculate other properties for the best iteration. The following code lets me get the "accuracy" for each fold on the best iteration:

`pd.DataFrame(
    [svc_search.cv_results_[f"split{x}_test_score"] for x in range(cv)],
    index=[f"split{x}_test_score" for x in range(cv)],
).T.iloc[svc_search.best_index_]`

Like so:
`split0_test_score    0.500000
split1_test_score    0.875000
split2_test_score    0.500000
split3_test_score    0.625000
split4_test_score    0.625000
split5_test_score    0.625000
split6_test_score    0.875000
split7_test_score    0.875000
split8_test_score    0.875000
split9_test_score    0.714286`

It would be really nice if I was able to calculate other metrics, such as sensitivity, specificity, Cohen's kappa, etc. I hope this makes sense.

**Comment by kernc** [comment]
Passing `scoring: dict =` and `refit: str =` as in `sklearn.model_selection.GridSearchCV` sounds reasonable, and at first glance looks mostly a documentation issue with but a few aligning touches needed.

Ideally though, the PR implementer would investigate the relevant new solutions in scikit-learn since 2017 (https://github.com/scikit-learn/scikit-learn/issues/9499) and have a look at and try to remove as much of the surrounding code in favor of deferring to methods of the parent class (or `RandomSearchCV`, whichever has a stable API interface), which handles all those already. @pietermarsman, if you're still interested.
